File: bytedance2019/5.py
n = int(input())
matrix = []
for i in range(n):
    matrix.append(list(map(int, input().split(' '))))

V = 1 << (n - 1)
dp = [[float('inf')] * V for i in range(n)]
for i in range(n):
    dp[i][0] = matrix[i][0]

for j in range(1, V):
    for i in range(n):
        if (i < 1 and ((j << (1 - i)) & 1) == 0) or ((i >= 1 and (j >> (i - 1)) & 1) == 0):
            for k in range(1, n):
                if ((j >> (k - 1)) & 1) == 1:
                    dp[i][j] = min(dp[i][j], matrix[i][k] + dp[k][j ^ (1 << (k - 1))])
print(dp[0][V - 1])

File: bytedance2019/5.java
package main.nowcoder.bytedance2019;

import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[][] matrix = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                matrix[i][j] = in.nextInt();
            }
        }
        in.close();

        int V = 1 << (n - 1);
        int[][] dp = new int[n][V];
        for (int i = 0; i < n; i++) {
            dp[i][0] = matrix[i][0];
        }

        for (int j = 1; j < V; j++) {
            for (int i = 0; i < n; i++) {
                dp[i][j] = Integer.MAX_VALUE;
                if (((j >> (i - 1)) & 1) == 0) {
                    for (int k = 1; k < n; k++) {
                        if (((j >> (k - 1)) & 1) == 1) {
                            dp[i][j] = Math.min(dp[i][j], matrix[i][k] + dp[k][j ^ (1 << (k - 1))]);
                        }
                    }
                }
            }
        }

        System.out.println(dp[0][V - 1]);
    }
}